Detectors/TRD/workflow/src/TRDTrackletTransformerSpec.cxx

```diff
@@ -27,45 +28,108 @@ namespace trd
+        const auto& trigRec = trigRecs[j];
+        if (trigRec.getBCData() >= irFrame.getMin()) {
+          if (trigRec.getBCData() <= irFrame.getMax()) {
+            // TRD interaction record inside ITS frame
+            trigRecBitfield[j] = 1;
+            lastMatchedIdx = j;
+          } else {
+            // too late, also the higher trigger records won't match
+            break;
+          }
+        }
+      }
+      LOGF(DEBUG, "ITS IR Frame start: %li, end: %li", irFrame.getMin().toLong(), irFrame.getMax().toLong());
+    }
+    /*
+    // for debugging: print TRD trigger times which are accepted and which are filtered out
+    for (int j = 0; j < trigRecs.size(); ++j) {
+      const auto& trigRec = trigRecs[j];
+      if (!trigRecBitfield[j]) {
+        LOGF(DEBUG, "Could not find ITS info for TRD trigger %i: %li", j, trigRec.getBCData().toLong());
+      } else {
+        LOGF(DEBUG, "Found ITS info for TRD trigger %i: %li", j, trigRec.getBCData().toLong());
+      }
+    }
+    */
+  } else {
+    // fill bitmask with 1
+    std::fill(trigRecBitfield.begin(), trigRecBitfield.end(), 1);
+  }
+
+  if (mTrigRecFilterActive) {
+    // skip tracklets from TRD triggers without ITS data
+    for (int iTrig = 0; iTrig < trigRecs.size(); ++iTrig) {
+      if (!trigRecBitfield[iTrig]) {
+        continue;
+      } else {
+        const auto& trigRec = trigRecs[iTrig];
+        for (int iTrklt = trigRec.getFirstTracklet(); iTrklt < trigRec.getFirstTracklet() + trigRec.getNumberOfTracklets(); ++iTrklt) {
+          calibratedTracklets[iTrklt] = mTransformer.transformTracklet(tracklets[iTrklt]);
+          ++nTrackletsTransformed;
+        }
+      }
+    }
+  } else {
+    // transform all tracklets
+    for (int iTrklt = 0; iTrklt < tracklets.size(); ++iTrklt) {
+      calibratedTracklets[iTrklt] = mTransformer.transformTracklet(tracklets[iTrklt]);
+      ++nTrackletsTransformed;
+    }
   }
 
+  LOGF(INFO, "Found %lu tracklets. Applied filter for ITS IR frames: %i. Transformed %i tracklets.", tracklets.size(), mTrigRecFilterActive, nTrackletsTransformed);
+
   pc.outputs().snapshot(Output{"TRD", "CTRACKLETS", 0, Lifetime::Timeframe}, calibratedTracklets);
+  pc.outputs().snapshot(Output{"TRD", "TRIGRECMASK", 0, Lifetime::Timeframe}, trigRecBitfield);
 }
 
-o2::framework::DataProcessorSpec getTRDTrackletTransformerSpec()
+o2::framework::DataProcessorSpec getTRDTrackletTransformerSpec(bool trigRecFilterActive)
 {
-  LOG(info) << "getting TRDTrackletTransformerSpec";
+  std::shared_ptr<DataRequest> dataRequest = std::make_shared<DataRequest>();
+  if (trigRecFilterActive) {
+    dataRequest->requestIRFramesITS();
+  }
+  auto& inputs = dataRequest->inputs;
+  inputs.emplace_back("trdtracklets", "TRD", "TRACKLETS", 0, Lifetime::Timeframe);
+  inputs.emplace_back("trdtriggerrec", "TRD", "TRKTRGRD", 0, Lifetime::Timeframe);
+
+  std::vector<OutputSpec> outputs;
+  outputs.emplace_back("TRD", "CTRACKLETS", 0, Lifetime::Timeframe);
```

Comment:
I would propose to set the subspec to 1 in case we apply the filtering, and then the tracker would request subspec 1 in case the filtering is active. That would ensure consistency.

Reply:
There is also the `TRIGRECMASK` in the line below. This is always filled by the tracklet transformer and taken into account by the tracking. This is a bitmap flagging the TRD triggers which are ignored/used. I think that avoids inconsistencys.
On the other hand I want to warn the user in case the filtering is active, but matching to TPC-only tracks is requested (https://github.com/AliceO2Group/AliceO2/pull/6585/files#diff-70a51291b8fcbd8a6f7ee3ffcfb34e8295217d05bedb09a2d16a15b869c9a0e1R264). For that the tracker needs to know whether the filtering was active or not. So I think I will do as you suggest and force the user to set the option both for the tracklet transformer and for the tracking and use the subspec.

Reply:
In principle, the matching to TPC-only should be reserved to async mode, where there is no event filtering anyway.
But I did not understand how using subspec = 1 improves consistency (what problem does this solve?)